lib/SILGen/SILGenApply.cpp

```diff
@@ -1053,7 +1054,18 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
     auto constant = SILDeclRef(afd).asForeign(requiresForeignEntryPoint(afd));
 
     auto subs = e->getDeclRef().getSubstitutions();
-    setCallee(Callee::forClassMethod(SGF, constant, subs, e));
+
+    bool isObjCDirect = false;
+    if (auto objcDecl = dyn_cast_or_null<clang::ObjCMethodDecl>(
+            afd->getClangDecl())) {
+      isObjCDirect = objcDecl->isDirectMethod();
```

Comment:
Can you add a test for Swift protocol conformances where the requirement is satisfied by an `objc_direct` witness? Emission of the call to the witness from the body of a witness thunk doesn't go through SILGenApply. I think we might need to insert a similar check in getWitnessDispatchKind() or thereabouts.

Reply:
I ran into this error https://github.com/llvm/llvm-project/blob/20d704a75ed51c7a9a155aa3978d0c02671c3f69/clang/test/SemaObjC/method-direct.m#L64 while trying to add this test. Since ObjC doesn't allow `objc_direct` on protocols, doesn't that make this code pattern impossible?

Reply:
I think you need a Swift protocol and ObjC class that implements that protocol and where the witness is annotated with `__attribute__((objc_direct))`.

Reply:
It shouldn't matter whether the original protocol is in Swift for ObjC. When clang runs, it will see an ObjC protocol and disallow any `objc_direct` methods on that protocol.

Reply:
Both the protocol and the conformance should be defined in Swift. Eg,

```
protocol MyProto {
  func myMethod()
}

extension MyObjCClass : MyProto {}
```

Then MyObjCClass would have an `objc_direct` method named `myMethod()`.